web/shapeworks/src/store/methods.ts

```diff
@@ -392,47 +392,232 @@ export async function cacheAllComparisons(comparisons: CacheComparison[][]) {
+                        )
+                        return [domain, domainRecords]
+                    })
+                )]
+            })
         )
-        if (subjectParticles.length > 0){
-            const numRows = Math.max(...subjectParticles)
-            landmarkInfo.value = [...Array(numRows).keys()].map((index) => {
-                let currentInfo = {
-                    id: index,
-                    color: COLORS[index % COLORS.length],
-                    name: `L${index}`,
-                    num_set: subjectParticles.filter(
-                        (numLocations) => numLocations > index
-                    ).length,
-                    comment: undefined,
+    }
+}
 
-                }
-                if (selectedProject.value?.landmarks_info && selectedProject.value.landmarks_info.length > index) {
-                    currentInfo = Object.assign(
-                        currentInfo,
-                        selectedProject.value?.landmarks_info[index]
-                    )
-                }
-                if (currentInfo.color.toString().includes("#")) {
-                    currentInfo.color = hexToRgb(currentInfo.color.toString())
-                }
-                return currentInfo
+export function reassignLandmarkNumSetValues() {
+    // only reassign if a value has changed
+    const reassign = false
+    const newInfo = landmarkInfo.value.map((lInfo) => {
+        const newNumSet = allSubjectsForDataset.value.filter((s) => !!getLandmarkLocation(s, lInfo)).length
+        if (newNumSet != lInfo.num_set) {
+            lInfo.num_set = newNumSet
+        }
+        return lInfo
+    })
+    if (reassign) landmarkInfo.value = newInfo
+}
+
+
+export async function getLandmarks() {
+    allSetLandmarks.value = {}
+    if(selectedProject.value?.landmarks) {
+        landmarkInfo.value = selectedProject.value.landmarks_info.map((lInfo, index) => {
+            if (!lInfo.id) {
+                lInfo.id = index
+            }
+            if (!lInfo.name) {
+                lInfo.name = `L${index}`
+            }
+            if (!lInfo.color) {
+                lInfo.color = COLORS[index % COLORS.length]
+            }
+            if (lInfo.color.toString().includes('#')) {
+                lInfo.color = hexToRgb(lInfo.color.toString())
+            }
+            lInfo.num_set = 0
+            return lInfo
+        })
+
+        for (let i = 0; i < selectedProject.value.landmarks.length; i++) {
+            const landmarksObject = selectedProject.value.landmarks[i]
+            const subject = allSubjectsForDataset.value.find((s) => s.id === landmarksObject.subject)
+            const pointData = await pointsReader(landmarksObject.file)
+            const locationData = pointData.getPoints().getData()
+            const locations: number[][] = []
+            for (let p = 0; p < locationData.length; p+=3) {
+                const location = locationData.slice(p, p+3) as number[]
```

Comment:
What is the reasoning behind `p+=3` and `p+3`? Can the `3` value ever change or is that a constant for landmarks? I believe this is the 2nd place I've seen that.

Reply:
The value for `pointData.getPoints().getData()` is actually a flattened array of points, like this: `[x1, y1, z1, x2, y2, z2...]` In order to iterate over the points in this list, we use a value `p` which increases by 3 on each iteration and we get a slice of the array from `p` to `p+3`. We can do this because we know each point must have three components.